### PR TITLE
[SPARK-43107][SQL] Coalesce buckets in join applied on broadcast join stream side

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This PR adds support coalesce buckets in join applied on broadcast join stream side.

### Why are the changes needed?

Reduce shuffle to improve query performance.

Before | After
-- | --
 | 

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/bucketing/CoalesceBucketsInJoin.scala [hunk 1]
```diff
@@ -115,7 +115,11 @@ object ExtractJoinWithBuckets {
   private def hasScanOperation(plan: SparkPlan): Boolean = plan match {
     case f: FilterExec => hasScanOperation(f.child)
     case p: ProjectExec => hasScanOperation(p.child)
-    case _: FileSourceScanExec => true
+    case j: BroadcastHashJoinExec =>
+      if (j.buildSide == BuildLeft) hasScanOperation(j.right) else hasScanOperation(j.left)
+    case j: BroadcastNestedLoopJoinExec =>
+      if (j.buildSide == BuildLeft) hasScanOperation(j.right) else hasScanOperation(j.left)
+    case f: FileSourceScanExec => f.relation.bucketSpec.nonEmpty
```
Comment (wangyum) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/bucketing/CoalesceBucketsInJoin.scala:122]: This optimization for bucket only, exclude non-bucket relation to reduce the following checks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/bucketing/CoalesceBucketsInJoin.scala [hunk 2]
```diff
@@ -142,9 +146,7 @@ object ExtractJoinWithBuckets {
   }
 
   private def isApplicable(j: ShuffledJoin): Boolean = {
-    (j.isInstanceOf[SortMergeJoinExec] ||
-      j.isInstanceOf[ShuffledHashJoinExec]) &&
```
Comment (wangyum) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/bucketing/CoalesceBucketsInJoin.scala:146]: It must be `ShuffledJoin`. These 2 check is not needed.

**Comment by wangyum** [comment]
cc @cloud-fan

**Review by cloud-fan — APPROVED** [review]

**Comment by wangyum** [comment]
Merged to master.

**Pull request closed without merging**